### PR TITLE
Friends don't let firends use String.format in logging statements

### PR DESCRIPTION
**Pull request by stephenc**
See https://javaadventure.blogspot.ie/2016/12/the-costs-of-java-utils-logging.html for a benchmark that confirms the cost

@i386 @batmat

### CODE REVIEW

**Comment by stephenc** [comment]
@batmat would you believe that the test failure was from the log messages not being parsed correctly too!

**Comment by michaelneale** [comment]
is there a JIRA ticket for this? As this may affect a few areas/problems (always best to have a ticket as people will chew out when there isn't one).

**Comment by michaelneale** [comment]
and ROFLscale blog post is not a valid ticket - I assume this broke something?

**Comment by michaelneale** [comment]
ah - no the bug I was thinking of: https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/83 

never mind.. .carry on. Nice one.

File: src/test/java/com/cloudbees/jenkins/plugins/bitbucket/WebhooksAutoregisterTest.java [hunk 1]
```diff
@@ -102,7 +103,11 @@ private void waitForLogFileMessage(String string, RingBufferLogHandler logs) thr
             while (limit < 5) {
                 rootDir.wait(1000);
                 for (LogRecord r : logs.getView()) {
-                    if (r.getMessage().contains(string)) {
+                    String message = r.getMessage();
+                    if (r.getParameters() != null) {
+                        message = MessageFormat.format(message, r.getParameters());
+                    }
+                    if (message.contains(string)) {
```
Comment (jglick) [inline src/test/java/com/cloudbees/jenkins/plugins/bitbucket/WebhooksAutoregisterTest.java]: use `LoggerRule`

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/WebhookAutoRegisterListener.java [hunk 2]
```diff
@@ -215,11 +215,12 @@ private void removeHooks(SCMSourceOwner owner) throws IOException, InterruptedEx
                     }
                 }
                 if (hook != null && !isUsedSomewhereElse(owner, source.getRepoOwner(), source.getRepository())) {
-                    LOGGER.info(String.format("Removing hook for %s/%s", source.getRepoOwner(), source.getRepository()));
+                    LOGGER.log(Level.INFO, "Removing hook for {0}/{1}",
+                            new Object[]{source.getRepoOwner(), source.getRepository()});
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/WebhookAutoRegisterListener.java]: irrelevant for `INFO` and above, which are almost certainly logged anyway, but harmless

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 3]
```diff
@@ -302,7 +303,8 @@ public String getDefaultBranch() throws IOException, InterruptedException {
         try {
             response = getRequest(url);
         } catch (FileNotFoundException e) {
-            LOGGER.fine(String.format("Could not find default branch for %s/%s", this.owner, this.repositoryName));
+            LOGGER.log(Level.FINE, "Could not find default branch for {0}/{1}",
+                    new Object[]{this.owner, this.repositoryName});
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: or more easily, in Java 8 just prepend `() -> ` to the message

Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Not on Java 8 baseline yet, and iirc JMH shows the array as faster than INDY

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/WebhookAutoRegisterListener.java (hunk 2 above)
Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/WebhookAutoRegisterListener.java]: JMH shows this form as faster than String.format. 

Now simple string concatenation is faster still for those cases where the message is logged, so we could argue that string concatenation should be used from WARNING up (as some - not many - users turn off INFO)

**Pull request merged**